main/src/ca/uwaterloo/flix/api/Flix.scala

```diff
@@ -441,7 +441,7 @@ class Flix {
   /**
     * Compiles the Flix program and returns a typed ast.
     */
-  def check(): Validation[TypedAst.Root, CompilationMessage] = {
+  def check(): Validation[TypedAst.Root, CompilationMessage] = try {
```

Comment:
A little unfortunate to introduce a try-catch-rethrow as we're working toward making things more functional. Could we make this a subtype of ComplilationMessage? I think that would lend itself better to LSP integration, though it would likely complicate things.

Reply:
Unfortunately not. The reason is that most of the backend compiler phase do not use Validation-- Why would they? They are not supposed to find errors. However when an abnormal situation is encountered they throw exceptions. 

You could argue against this approach, and maybe in the future that could be different, but for now there is a lot of code in that style.